File: extract-modules.py
#!/usr/bin/env python3
"""
Estrae la colonna J (MODULI ENVIZI, COME E BENEFICI) da
pscenari-moduli-referenze.xlsx e produce SCENARIO_MODULES per constants.ts.

Uso: python3 envizi-quest-v3/extract-modules.py
"""
import json
import re
import openpyxl

XLSX = "envizi-quest-v3/source/public/pscenari-moduli-referenze.xlsx"
CONSTANTS = "envizi-quest-v3/source/src/constants.ts"

wb = openpyxl.load_workbook(XLSX, data_only=True)
ws = wb["Voci Esaminate"]

# Legge tutte le righe dati (riga 2..211), colonna J (indice 10, 1-based)
col_j_values = []
for row in ws.iter_rows(min_row=2, max_row=211, min_col=10, max_col=10, values_only=True):
    val = row[0]
    # Normalizza newline: sostituisce \n con \\n per compatibilità TypeScript
    raw = str(val).strip() if val is not None else ""
    col_j_values.append(raw.replace("\n", "\\n"))

print(f"# Righe lette: {len(col_j_values)}", flush=True)

# Raggruppa ogni 5 righe consecutive = un needId (42 gruppi totali)
GROUP_SIZE = 5
groups = []
for i in range(0, len(col_j_values), GROUP_SIZE):
    groups.append(col_j_values[i:i+GROUP_SIZE])

print(f"# Gruppi: {len(groups)}", flush=True)

# Legge i needId in ordine da constants.ts
with open(CONSTANTS, "r", encoding="utf-8") as f:
    constants_text = f.read()

needle_ids = re.findall(r'"([a-z]+-\d+)"\s*:', constants_text)
# Deduplica mantenendo l'ordine
seen = set()
need_ids = []
for nid in needle_ids:
    if nid not in seen:
        seen.add(nid)
        need_ids.append(nid)

print(f"# needIds: {len(need_ids)}", flush=True)

if len(need_ids) != len(groups):
    print(f"# ATTENZIONE: {len(need_ids)} needIds vs {len(groups)} gruppi")

# Genera la costante
lines = ["export const SCENARIO_MODULES:Record<string,string[]>={"]
for i, group in enumerate(groups):
    nid = need_ids[i] if i < len(need_ids) else f"unknown-{i}"
    vals = json.dumps(group, ensure_ascii=False)
    # Toglie le parentesi quadre esterne per inserirle nel template
    lines.append(f"  {json.dumps(nid)}:{vals},")
lines.append("};")

print("\n".join(lines))

File: extract-modules.mjs
/**
 * Estrae i valori della colonna J (MODULI ENVIZI, COME E BENEFICI) da
 * pscenari-moduli-referenze.xlsx e produce la costante SCENARIO_MODULES
 * da incollare in constants.ts.
 *
 * Eseguire con:  node extract-modules.mjs
 */
import { readFile } from "node:fs/promises";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dir = dirname(fileURLToPath(import.meta.url));

// Legge xlsx tramite il modulo già installato nel progetto
const xlsxPath = join(__dir, "source/node_modules/xlsx");
const XLSX = await import(xlsxPath + "/xlsx.mjs").catch(() => import(xlsxPath));

const filePath = join(__dir, "source/public/pscenari-moduli-referenze.xlsx");
const buf = await readFile(filePath);
const wb = XLSX.read(buf, { type: "buffer" });
const ws = wb.Sheets["Voci Esaminate"];
const rows = XLSX.utils.sheet_to_json(ws, { header: 1, defval: "" });

// Colonna B = indice 1 (Voce Esaminata / needId base)
// Colonna I = indice 8 (SCENARIO NARRATIVO — per verifica)
// Colonna J = indice 9 (MODULI ENVIZI, COME E BENEFICI)
// Le righe dati partono da indice 1 (riga 2 Excel)

// Mappa dal testo needId atteso ai dati raggruppati
// Ogni needId ha 5 scenari consecutivi; la voce Esaminata è la stessa per il gruppo
// Costruiamo la mappa: voceName -> array di 5 valori col J (nell'ordine delle righe)

const groups = new Map(); // voceName -> [colJ, ...]

for (let i = 1; i < rows.length; i++) {
  const row = rows[i];
  const voce = String(row[1] ?? "").trim();
  const colJ = String(row[9] ?? "").trim();
  if (!voce) continue;
  if (!groups.has(voce)) groups.set(voce, []);
  groups.get(voce).push(colJ);
}

// Legge USE_CASE_SCENARIOS da constants.ts per mappare voce→needId
const constantsPath = join(__dir, "source/src/constants.ts");
const { readFileSync } = await import("node:fs");
const constantsText = readFileSync(constantsPath, "utf8");

// Estrae le chiavi di USE_CASE_SCENARIOS
const keysMatch = constantsText.match(/USE_CASE_SCENARIOS\s*:\s*Record<[^>]+>\s*=\s*\{([\s\S]*?)^\}/m);

// Approccio semplice: legge la stessa colonna B per costruire il mapping needId
// Poiché USE_CASE_SCENARIOS usa chiavi come "credit-1", dobbiamo estrarre le chiavi direttamente
// Estraiamo tutte le chiavi di USE_CASE_SCENARIOS
const keyRegex = /"([a-z]+-\d+)"\s*:/g;
const needIds = [];
let m;
while ((m = keyRegex.exec(constantsText)) !== null) {
  // Evita duplicati
  if (!needIds.includes(m[1])) needIds.push(m[1]);
}

// Costruisce SCENARIO_MODULES: per ogni needId prende i 5 valori colJ dal gruppo corrispondente
// L'ordine delle voci deve corrispondere all'ordine in USE_CASE_SCENARIOS
// Leggiamo l'ordine delle voci dall'Excel (ogni 5 righe = un needId)
const voceOrder = [];
const seen = new Set();
for (let i = 1; i < rows.length; i++) {
  const voce = String(rows[i][1] ?? "").trim();
  if (voce && !seen.has(voce)) { seen.add(voce); voceOrder.push(voce); }
}

console.log("// SCENARIO_MODULES generato da pscenari-moduli-referenze.xlsx — colonna J");
console.log("export const SCENARIO_MODULES:Record<string,string[]>={");

voceOrder.forEach((voce, idx) => {
  const needId = needIds[idx];
  if (!needId) { console.error(`MANCANTE needId per voce[${idx}]: ${voce}`); return; }
  const vals = groups.get(voce) ?? [];
  const escaped = vals.map(v => JSON.stringify(v));
  console.log(`  ${JSON.stringify(needId)}:[${escaped.join(",")}],`);
});

console.log("};");
